Assets/MixedRealityToolkit.Providers/XRSDK/XRSDKDeviceManager.cs

```diff
@@ -0,0 +1,244 @@
+
+using Microsoft.MixedReality.Toolkit.Input;
+using Microsoft.MixedReality.Toolkit.Utilities;
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.XR;
+
+namespace Microsoft.MixedReality.Toolkit.XRSDK.Input
+{
+    /// <summary>
+    /// Manages XR SDK devices.
+    /// </summary>
+    [MixedRealityDataProvider(
+        typeof(IMixedRealityInputSystem),
+        SupportedPlatforms.WindowsStandalone | SupportedPlatforms.MacStandalone | SupportedPlatforms.LinuxStandalone | SupportedPlatforms.WindowsUniversal,
+        "XRSDK Device Manager")]
+    public class XRSDKDeviceManager : BaseInputDeviceManager, IMixedRealityCapabilityCheck
+    {
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="inputSystem">The <see cref="Microsoft.MixedReality.Toolkit.Input.IMixedRealityInputSystem"/> instance that receives data from this provider.</param>
+        /// <param name="name">Friendly name of the service.</param>
+        /// <param name="priority">Service priority. Used to determine order of instantiation.</param>
+        /// <param name="profile">The service's configuration profile.</param>
+        public XRSDKDeviceManager(
+            IMixedRealityInputSystem inputSystem,
+            string name = null,
+            uint priority = DefaultPriority,
+            BaseMixedRealityProfile profile = null) : base(inputSystem, name, priority, profile) { }
+
+        /// <inheritdoc />
+        public virtual bool CheckCapability(MixedRealityCapability capability)
+        {
+            // The XR SDK platform supports motion controllers.
+            return (capability == MixedRealityCapability.MotionController);
+        }
+
+        protected static readonly Dictionary<string, GenericXRSDKController> ActiveControllers = new Dictionary<string, GenericXRSDKController>();
+
+        private InputDevice leftInputDevice;
+        private InputDevice rightInputDevice;
+        private bool wasLeftInputDeviceValid = false;
+        private bool wasRightInputDeviceValid = false;
+
+        /// <inheritdoc/>
+        public override void Update()
+        {
+            base.Update();
+
+            if (XRSDKSubsystemHelpers.InputSubsystem == null || !XRSDKSubsystemHelpers.InputSubsystem.running)
+            {
+                return;
+            }
+
+            if (!leftInputDevice.isValid)
+            {
+                if (wasLeftInputDeviceValid)
+                {
+                    GenericXRSDKController controller = GetOrAddController(leftInputDevice);
+
+                    if (controller != null)
+                    {
+                        CoreServices.InputSystem?.RaiseSourceLost(controller.InputSource, controller);
+                    }
+
+                    RemoveController(leftInputDevice);
+                    wasLeftInputDeviceValid = false;
+                }
+
+                leftInputDevice = InputDevices.GetDeviceAtXRNode(XRNode.LeftHand);
```

Comment:
>GetDeviceAtXRNode [](start = 47, length = 17)

This transition from invalid->valid->invalid happens when the hand loses tracking right? It's kinda interesting wrapping my head around seeing an InputDevice for a left hand being permanently sunset (because generally people tend to have the same left hand in the same session i.e. their left hand)